ako-gateway-api/k8s/gateway_controller.go

```diff
@@ -0,0 +1,585 @@
+					}
+					if lib.UseServicesAPI() {
+						utils.AviLog.Debugf("Deleting Gatways for namespaces: %s", nsCur.GetName())
+						AddGatewaysFromNSToIngestionQueue(numWorkers, c, nsCur.GetName(), lib.NsFilterDelete)
+					}
+				}
+			}
+		},
+	}
+	return namespaceEventHandler
+}
+func isNamespaceUpdated(oldNS, newNS *corev1.Namespace) bool {
+	if oldNS.ResourceVersion == newNS.ResourceVersion {
+		return false
+	}
+	oldLabelHash := utils.Hash(utils.Stringify(oldNS.Labels))
+	newLabelHash := utils.Hash(utils.Stringify(newNS.Labels))
+	if oldLabelHash != newLabelHash {
+		return true
+	}
+	return false
+}
+
+func AddServicesFromNSToIngestionQueue(numWorkers uint32, c *GatewayController, namespace string, msg string) {
+	var key string
+	svcObjs, err := utils.GetInformers().ServiceInformer.Lister().Services(namespace).List(labels.Set(nil).AsSelector())
+	if err != nil {
+		utils.AviLog.Errorf("Unable to retrieve the services during namespace sync: %s", err)
+		return
+	}
+	for _, svcObj := range svcObjs {
+
+		key = utils.Service + "/" + utils.ObjKey(svcObj)
+
+		bkt := utils.Bkt(namespace, numWorkers)
+		c.workqueue[bkt].AddRateLimited(key)
+		utils.AviLog.Debugf("key: %s, msg: %s for namespace: %s", key, msg, namespace)
+	}
+}
+
+func AddGatewaysFromNSToIngestionQueue(numWorkers uint32, c *GatewayController, namespace string, msg string) {
+	gwObjs, err := gwlib.AKOControlConfig().GatewayApiInformers().GatewayInformer.Lister().Gateways(namespace).List(labels.Set(nil).AsSelector())
+	if err != nil {
+		utils.AviLog.Errorf("NS to gateways queue add: Error occurred while retrieving gateways for namespace: %s", namespace)
+		return
+	}
+	for _, gwObj := range gwObjs {
+		key := utils.Gateway + "/" + utils.ObjKey(gwObj)
+		bkt := utils.Bkt(namespace, numWorkers)
+		c.workqueue[bkt].AddRateLimited(key)
+		utils.AviLog.Debugf("key: %s, msg: %s for namespace: %s", key, msg, namespace)
+	}
+
+	gwClassObjs, err := gwlib.AKOControlConfig().GatewayApiInformers().GatewayClassInformer.Lister().List(labels.Set(nil).AsSelector())
+	if err != nil {
+		utils.AviLog.Errorf("NS to gateways queue add: Error occurred while retrieving gateway class for namespace: %s", namespace)
+		return
+	}
+	for _, gwClassObj := range gwClassObjs {
+		key := utils.GatewayClass + "/" + utils.ObjKey(gwClassObj)
+		bkt := utils.Bkt(namespace, numWorkers)
+		c.workqueue[bkt].AddRateLimited(key)
+		utils.AviLog.Debugf("key: %s, msg: %s for namespace: %s", key, msg, namespace)
+	}
+}
+
+func checkAviSecretUpdateAndShutdown(secret *corev1.Secret) bool {
+	if secret.Namespace == utils.GetAKONamespace() && secret.Name == lib.AviSecret {
+		// if the secret is updated or deleted we shutdown API server
+		utils.AviLog.Warnf("Avi Secret object %s/%s updated/deleted, shutting down AKO", secret.Namespace, secret.Name)
+		lib.ShutdownApi()
+		return false
+	}
+	return true
+}
+
+func (c *GatewayController) SetupGatewayApiEventHandlers(numWorkers uint32) {
```

Comment:
Shall we move this to a separate file that contains all the event handlers for better code browsing?